drivers/docker/handle.go

```diff
@@ -0,0 +1,292 @@
+		OutputStream: stdout,
+		ErrorStream:  stderr,
+		Context:      ctx,
+	}
+	if err := client.StartExec(exec.ID, startOpts); err != nil {
+		return nil, err
+	}
+	execResult.Stdout = stdout.Bytes()
+	execResult.Stderr = stderr.Bytes()
+	res, err := client.InspectExec(exec.ID)
+	if err != nil {
+		return execResult, err
+	}
+
+	execResult.ExitResult.ExitCode = res.ExitCode
+	return execResult, nil
+}
+
+func (h *taskHandle) Signal(s os.Signal) error {
+	// Convert types
+	sysSig, ok := s.(syscall.Signal)
+	if !ok {
+		return fmt.Errorf("Failed to determine signal number")
+	}
+
+	// TODO When we expose signals we will need a mapping layer that converts
+	// MacOS signals to the correct signal number for docker. Or we change the
+	// interface to take a signal string and leave it up to driver to map?
+
+	dockerSignal := docker.Signal(sysSig)
+	opts := docker.KillContainerOptions{
+		ID:     h.container.ID,
+		Signal: dockerSignal,
+	}
+	return h.client.KillContainer(opts)
+
+}
+
+// Kill is used to terminate the task.
+func (h *taskHandle) Kill(killTimeout time.Duration, signal os.Signal) error {
+	// Only send signal if killTimeout is set, otherwise stop container
+	if killTimeout > 0 {
+		if err := h.Signal(signal); err != nil {
+			return err
+		}
+		select {
+		case <-h.waitCh:
+			return nil
+		case <-time.After(killTimeout):
+		}
+	}
+
+	// Stop the container
+	err := h.client.StopContainer(h.container.ID, 0)
+	if err != nil {
+
+		// Container has already been removed.
+		if strings.Contains(err.Error(), NoSuchContainerError) {
+			h.logger.Debug("attempted to stop nonexistent container")
+			return nil
+		}
+		h.logger.Error("failed to stop container", "error", err)
+		return fmt.Errorf("Failed to stop container %s: %s", h.container.ID, err)
+	}
+	h.logger.Info("stopped container")
+	return nil
+}
+
+func (h *taskHandle) Stats() (*structs.TaskResourceUsage, error) {
+	h.resourceUsageLock.RLock()
+	defer h.resourceUsageLock.RUnlock()
+	var err error
+	if h.resourceUsage == nil {
+		err = fmt.Errorf("stats collection hasn't started yet")
+	}
+	return h.resourceUsage, err
+}
+
+func (h *taskHandle) run() {
+	exitCode, werr := h.waitClient.WaitContainer(h.container.ID)
+	if werr != nil {
+		h.logger.Error("failed to wait for container; already terminated")
+	}
+
+	if exitCode != 0 {
+		werr = fmt.Errorf("Docker container exited with non-zero exit code: %d", exitCode)
+	}
+
+	container, ierr := h.waitClient.InspectContainer(h.container.ID)
+	oom := false
+	if ierr != nil {
+		h.logger.Error("failed to inspect container", "error", ierr)
+	} else if container.State.OOMKilled {
+		oom = true
+		werr = fmt.Errorf("OOM Killed")
+		labels := []metrics.Label{
+			{
+				Name:  "job",
+				Value: h.task.JobName,
+			},
+			{
+				Name:  "task_group",
+				Value: h.task.TaskGroupName,
+			},
+			{
+				Name:  "task",
+				Value: h.task.Name,
+			},
+		}
+		metrics.IncrCounterWithLabels([]string{"driver", "docker", "oom"}, 1, labels)
```

Comment:
We should do this in the task runner since as an external binary the global `metrics` sink won't be properly configured (the agent configures it).